src/Microsoft.DocAsCode.Build.Engine/MarkdownProviders/jsonTokenTreeServiceProvider.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DocAsCode.Build.Engine
+{
+    using System.Collections.Generic;
+    using System.Collections.Immutable;
+    using System.Composition;
+    using System.Text;
+
+    using Microsoft.DocAsCode.Dfm;
+    using Microsoft.DocAsCode.MarkdownLite;
+    using Microsoft.DocAsCode.Plugins;
+
+    [Export("jsonTokenTree", typeof(IMarkdownServiceProvider))]
+    public class JsonTokenTreeServiceProvider : IMarkdownServiceProvider
+    {
+        public IMarkdownService CreateMarkdownService(MarkdownServiceParameters parameters)
+        {
+            return new JsonTokenTreeService(
+                parameters.BasePath,
+                parameters.Tokens,
+                MarkdownTokenTreeValidatorFactory.Combine(TokenTreeValidator));
+        }
+
+        [ImportMany]
+        public IEnumerable<IMarkdownTokenTreeValidator> TokenTreeValidator { get; set; }
+
+        private sealed class JsonTokenTreeService : IMarkdownService
+        {
+            private readonly DfmEngineBuilder _builder;
+
+            private readonly ImmutableDictionary<string, string> _tokens;
+
+            public JsonTokenTreeService(string baseDir, ImmutableDictionary<string, string> tokens, IMarkdownTokenTreeValidator tokenTreeValidator)
+            {
+                _builder = DocfxFlavoredMarked.CreateBuilder(baseDir);
+                _builder.TokenTreeValidator = tokenTreeValidator;
+                _tokens = tokens;
+            }
+
+            public MarkupResult Markup(string src, string path)
+            {
+                var dependency = new HashSet<string>();
+                var json = new StringBuilder(_builder.CreateDfmEngine(new JsonRenderer() { Tokens = _tokens }).Markup(src, path, dependency));
+                if (json.Length != 0)
+                {
+                    json.Insert(0, "{\"name\":\"markdown\",\"children\":[");
+                    json.Remove(json.Length - 1, 1);
+                    json.Append("]}");
+                }
```

Comment:
you  mean that I should avoid using the 'insert'? 

---
In reply to: [74869486](https://github.com/dotnet/docfx/pull/608#discussion_r74869486) [](ancestors = 74869486)